### DISCUSSION TITLE
Change languageId for particular file at runtime

### DISCUSSION THREAD

**Discussion by vaclavHala** [microsoft/vscode-discussions#1015] [Extension Development QnA] [answered]
We have some logic to determine arbitrary file in user workspace should be associated with our language. 
None of the builtin detection mechanisms on the `languages` contribution points (filenames, globs, first line etc.) can be used to satisfy our detection strategy.

We found two possible workarounds none of which we are very happy about:

**Associate language when file is opened**

```js
vscode.workspace.onDidOpenTextDocument(document => {
    if (isMagicalFile(document)) {
        vscode.languages.setTextDocumentLanguage(document, 'magic');
    }
})
```
There are several problems with this approach:
* file has wrong icon in explorer until it is opened
* icon reverts to bad when the `TextDocument` is closed again
* opening of the document is slowed down as changing language scratches the old `TextDocument` and starts opening from scratch with the new language

**Maintain association in workspace/user settings**

It is possible to have setting like this in workspace settings file:
```json
"settings": {
  "files.associations": {
    "/abs/path/to/some/file": "magic",
    "/abs/path/to/other/file": "magic"
  }
}
```
which we could keep up to date from our extension.

Downsides of this approach:
* we put garbage into users settings as we can't rely on present of workspace file which I feel is a big no-no
* it may become tedious to correctly update the shared list when user works with multiple workspaces, paths from which all get dumped into the single field
* we make it impossible for user to manually set association on their own in case the magic detection mechanism fails as their entry will now be replaced by our automated process keeping the list up to date

Are there some other ways to put the mapping into vscode I'm not seeing?

**Comment by abrilurena** [comment]
Hi, @vaclavHala! It's been a while, I just wanted to check in and see if you still need any help with this?

**Reply by vaclavHala** [reply]
Hi, yes please. I haven't managed to find any better solutions yet so I'd like some response giving better solution than those I already found, or saying there is currently no better way of going about this and I have to bite the bullet and pick one of the two above. Thanks

**Comment by theReynald** [comment]
Hi vaclavHala.

Given the constraints and considerations you've mentioned, besides the two workarounds you've identified, here are some alternative strategies to consider:

1. **Dynamic Language Configuration**: Instead of changing the language when the file is opened, consider delaying the language association until the user actively interacts with the file or performs a specific action. For example, you could provide a command or context menu option that allows the user to manually set the language for the current file. This way, the language association is only applied when needed, reducing the impact on file icons and performance.

2. **Temporary Language Association**: Rather than permanently changing the language of the file, you could temporarily associate the file with your language for the duration of the user session or until a specific condition is met. This temporary association could be implemented using a similar approach to the `onDidOpenTextDocument` event handler, but with the language reverted back once the file is closed or the condition is no longer met. This would avoid cluttering user settings with unnecessary associations.

3. **Virtual Documents**: Consider treating the files as "virtual" documents within your extension rather than changing their language association in the editor. You could implement your own language-specific features and functionality for these files within your extension's context, without affecting the editor's language settings or file icons. This approach allows you to maintain control over the language association without impacting the user's workspace configuration.

4. **Custom File Extension**: If possible, consider using a custom file extension for your files instead of relying on arbitrary file detection. By using a unique file extension specific to your language, you can easily associate these files with your language using the standard file association mechanisms provided by VS Code. This approach avoids the need for dynamic language association and reduces the risk of conflicts with other files in the workspace.

Ultimately, the best approach depends on the specific requirements and constraints of your extension and the user experience you want to provide but let me know if any of these help.

**Reply by vaclavHala** [reply]
Hi, thanks for the tips

ad 1 - this seems to me like just letting user pick file type via the language picker (bottom right status bar item) vscode provides, or am I missing something?

ad 2 - sounds like the `onDidOpenTextDocument` with addition of writing custom code to "revert" the language change immediately as the editor closes rather than waiting for vscode to revert it on its own at some point in the future. Don't think this helps me

ad 3 - not sure what you mean by virtual document, are we talking about writing custom `vscode.FileSystemProvider`? That sounds like more effort then I'm prepared to invest, also not sure it would work as that interface seems to work at the level of raw bytes (`readFile(uri: Uri): Uint8Array | Thenable<Uint8Array>;`), not document types, so I'd expect all the language detection features to work the same as if the file came directly from file system. Or did you have something completely different in mind?

ad 4 - yes that would be the best solution, sadly requirements given to me are such that extension is not under my control and can't be relied on

I have one more question regarding the association via 
```json
"settings": {
  "files.associations": {}
}
```
In large workspaces we sometimes get the dreaded "Unable to watch for file changes in this large workspace folder" errors. Is there some way we could nudge vscode to notice changes to the particular `settings.json` file after we modify it, to ensure updates written into it take effect even if regular file watching would miss it?

**Reply by burkeholland** [reply]
Using the vscode participant with /api, GitHub Copilot recommends this...

Create your own File System Watcher...

```typescript
const settingsPath = vscode.workspace.workspaceFolders
  ? path.join(vscode.workspace.workspaceFolders[0].uri.fsPath, '.vscode', 'settings.json')
  : null;

if (settingsPath) {
  const settingsWatcher = vscode.workspace.createFileSystemWatcher(settingsPath);

  settingsWatcher.onDidChange((e) => {
    console.log(`settings.json changed`);
    // Reload your extension, read new settings, etc.
  });

  settingsWatcher.onDidCreate((e) => {
    console.log(`settings.json created`);
    // Handle creation of settings.json
  });

  settingsWatcher.onDidDelete((e) => {
    console.log(`settings.json deleted`);
    // Handle deletion of settings.json
  });
}
```

**Comment by abrilurena** [comment]
Hi, @vaclavHala! i hope you're doing great. do you still need help with this?

**Reply by vaclavHala** [reply, marked as answer]
We ended up going with the `Maintain association in workspace/user settings` way. It seems to work well enough so I'll close this discussion

**Reply by abrilurena** [reply]
okidoki, thank you! please feel free to reach out if you need something else.
